src/path.cr

```diff
@@ -0,0 +1,836 @@
+  #
+  # ```
+  # Path["foo"].expand             # => Path["/current/path/foo"]
+  # Path["~/crystal/foo"].expand   # => Path["/home/crystal/foo"]
+  # Path["baz"].expand("/foo/bar") # => Path["/foo/bar/baz"]
+  # ```
+  def expand(base : Path | String = Path.current, home = Path.home) : Path
+    name = @name
+
+    if name.starts_with?('~')
+      home_ends_with_separator = @type.ends_with_separator?(home.@name)
+      home = new_instance(home.@name).normalize.to_s
+      # home = home + @type.separator if home_ends_with_separator && home.size > 1
+
+      if name.size == 1
+        return new_instance(home)
+      elsif @type.separators.includes?(name[1])
+        name = home + name[1..-1]
+      end
+    end
+
+    unless @type.starts_with_separator?(name)
+      base = new_instance(base.check_no_null_byte) unless base.is_a?(Path)
+      if name.empty?
+        name = base.expand.to_s
+      else
+        name = "#{base.expand}#{@type.separator}#{name}"
+      end
+    end
+
+    new_instance(name).normalize(remove_final_separator: false)
+  end
+
+  # Appends the given *parts* to this path and returns the joined path.
+  #
+  # ```
+  # Path["foo"].join("bar", "baz")       # => Path["foo/bar/baz"]
+  # Path["foo/"].join("/bar/", "/baz")   # => Path["foo/bar/baz"]
+  # Path["/foo/"].join("/bar/", "/baz/") # => Path["/foo/bar/baz/"]
+  # ```
+  def join(*parts) : Path
+    join parts
+  end
+
+  # Appends the given *parts* to this path and returns the joined path.
+  #
+  # ```
+  # Path["foo"].join("bar", "baz")           # => Path["foo/bar/baz"]
+  # Path["foo/"].join(Path["/bar/", "/baz"]) # => Path["foo/bar/baz"]
+  # Path["/foo/"].join("/bar/", "/baz/")     # => Path["/foo/bar/baz/"]
+  # ```
+  def join(parts : Array | Tuple) : Path
+    new_name = String.build do |str|
+      str << @name
+      last_ended_with_separator = @type.ends_with_separator?(@name)
+
+      parts.each_with_index do |part, index|
+        case part
+        when Path
+          part = part.@name
+        else
+          part = part.to_s
+          part.check_no_null_byte
+        end
+
+        byte_start = 0
+        byte_count = part.bytesize
+
+        case {@type.starts_with_separator?(part), last_ended_with_separator}
+        when {true, true}
+          byte_start += 1
+          byte_count -= 1
+        when {false, false}
+          str << @type.separator unless str.bytesize == 0
+        end
+
+        last_ended_with_separator = @type.ends_with_separator?(part)
+
+        str.write part.unsafe_byte_slice(byte_start, byte_count)
+      end
+    end
+
+    new_instance new_name
+  end
+
+  # Appends the given *part* to this path and returns the joined path.
+  #
+  # ```
+  # Path["foo"] / "bar" / "baz"     # => Path["foo/bar/baz"]
+  # Path["foo/"] / Path["/bar/baz"] # => Path["foo/bar/baz"]
+  # ```
+  def /(part : Path | String) : Path
+    join(part)
+  end
+
+  # :nodoc:
+  def <<(part : Path | String) : Path
```

Comment:
Please no, `Path` should be absolutely immutable.

Reply:
100% agree. This method is only required for array-like literals (`Path{"/foo/bar"}`). But as @asterite suggested, this should be changed.

Reply:
Second. Let path be inmutable.